File: test/instr/EnumTestRemoveGenerated.java
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.redextest;

import org.junit.Test;
import static org.fest.assertions.api.Assertions.assertThat;
import java.lang.reflect.Method;
import java.lang.reflect.InvocationTargetException;
import androidx.annotation.Nullable;
import java.io.Serializable;

enum UsesValueOf {
  ONE, TWO;
  public int otherField; /* prevent replace_enum_with_int */
}
enum UsesValuesMethod {
  ONE, TWO;
  public int otherField;
}
enum UsesNothing {
  ONE, TWO;
  public int otherField;
}
enum Captured {
  ONE, TWO;
  public int otherField;
}
enum UsedAsTypeClass {
  ONE, TWO;
  public int otherField;
}
enum Upcasted {
  ONE, TWO;
  public int otherField;
}
enum UpcastedToSerializable {
  ONE, TWO;
  public int otherField;
}

 /* Test cases */
public class EnumTestRemoveGenerated {

  private class CapturingClass {
      UsesValuesMethod s;
      Object obj;

      // `UsesValuesMethod` escapes as original type.
      public void capture_as_self(UsesValuesMethod arg) {
        this.s = arg;
      }

      // `Captured` escapes as object type.
      public void capture_as_object(Captured c) {
        this.obj = c;
      }
  };

  // `UsesValues` is upcasted and used as type `Class`, but it does not escape
  // the method.
  private boolean upcast_locally(UsesValuesMethod s) {
    switch ((int) (Math.random() * 2)) {
      case 0:
        return (Enum) s == null;
      default:
        return (Object) s == null;
    }
  }

  // Enums that are used as type `Class` could call one of many methods that
  // could lead to reflection, e.g., `Class.getMethods()`.
  private boolean use_as_type_class(UsedAsTypeClass s) {
    Class cl = s.getDeclaringClass();
    return cl == null;
  }

  // `Upcasted` is upcasted and escapes the method.
  public Enum return_type_enum(Upcasted s) {
    return s;
  }

  private boolean upcast_to_serializable(Serializable s) {
    return s == null;
  }

  @Test
  public void generated_methods() {
    CapturingClass capturingClass = new CapturingClass();
    capturingClass.capture_as_self(UsesValuesMethod.ONE);
    capturingClass.capture_as_object(Captured.ONE);
    upcast_locally(UsesValuesMethod.ONE);
    use_as_type_class(UsedAsTypeClass.ONE);
    return_type_enum(Upcasted.ONE);
    upcast_to_serializable(UpcastedToSerializable.ONE);

    UsesValueOf.valueOf("ONE");
    UsesValuesMethod.values();

    int a = UsesValueOf.ONE.ordinal();
    a = UsesValuesMethod.ONE.ordinal();
    a = UsesNothing.ONE.ordinal();
    a = Captured.TWO.ordinal();
    a = UsedAsTypeClass.ONE.ordinal();
    a = Upcasted.ONE.ordinal();
    a = UpcastedToSerializable.TWO.ordinal();
    a = UsesValuesMethod.ONE.compareTo(UsesValuesMethod.TWO);
    a = UsesValuesMethod.TWO.hashCode();
    boolean b = UsesValuesMethod.ONE.equals(UsesValuesMethod.TWO);
    String s = UsesValuesMethod.TWO.name();
    s = UsesValuesMethod.TWO.toString();
  }
}

File: test/instr/EnumTestRemoveGenerated.cpp
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <gtest/gtest.h>

#include "EnumClinitAnalysis.h"
#include "verify/VerifyUtil.h"

using namespace optimize_enums;

namespace {

const char* enum_usesvalueof_name = "Lcom/facebook/redextest/UsesValueOf;";
const char* enum_usesvaluesmethod_name =
    "Lcom/facebook/redextest/UsesValuesMethod;";
const char* enum_usesnothing_name = "Lcom/facebook/redextest/UsesNothing;";
const char* enum_captured_name = "Lcom/facebook/redextest/Captured;";
const char* enum_usedastypeclass_name =
    "Lcom/facebook/redextest/UsedAsTypeClass;";
const char* enum_upcasted_name = "Lcom/facebook/redextest/Upcasted;";
const char* enum_upcastedtoserializable_name =
    "Lcom/facebook/redextest/UpcastedToSerializable;";
const char* enum_instancefieldofserializable_name =
    "Lcom/facebook/redextest/InstanceFieldOfSerializable;";

bool has_valueof_method(DexType* enum_type) {
  auto proto = DexProto::get_proto(
      enum_type,
      DexTypeList::get_type_list({DexType::get_type("Ljava/lang/String;")}));
  return DexMethod::get_method(
             enum_type, DexString::make_string("valueOf"), proto) != nullptr;
}

bool has_values_method(DexType* enum_type) {
  auto proto = DexProto::get_proto(make_array_type(enum_type),
                                   DexTypeList::get_type_list({}));
  return DexMethod::get_method(
             enum_type, DexString::make_string("values"), proto) != nullptr;
}

bool has_values_field(DexType* enum_type) {
  return DexField::get_field(enum_type,
                             DexString::make_string("$VALUES"),
                             make_array_type(enum_type)) != nullptr;
}
} // namespace

/**
 * Test that the generated methods `valueOf()` and `values()` are removed if
 * safe.
 */
TEST_F(PostVerify, transform) {
  // UsesValueOf class
  auto enum_cls = find_class_named(classes, enum_usesvalueof_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // UsesValuesMethod class
  enum_cls = find_class_named(classes, enum_usesvaluesmethod_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  // EXPECT_FALSE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // UsesNothing class
  enum_cls = find_class_named(classes, enum_usesnothing_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  // EXPECT_FALSE(has_valueof_method(enum_cls->get_type()));
  // EXPECT_FALSE(has_values_method(enum_cls->get_type()));
  // EXPECT_FALSE(has_values_field(enum_cls->get_type()));

  // Captured class
  enum_cls = find_class_named(classes, enum_captured_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // UsedAsTypeClass class
  enum_cls = find_class_named(classes, enum_usedastypeclass_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // Upcasted class
  enum_cls = find_class_named(classes, enum_upcasted_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // UpcastedToSerializable class
  enum_cls = find_class_named(classes, enum_upcastedtoserializable_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));

  // InstanceOfSerializable class
  enum_cls = find_class_named(classes, enum_instancefieldofserializable_name);
  ASSERT_NE(enum_cls, nullptr);
  EXPECT_TRUE(is_enum(enum_cls));
  EXPECT_TRUE(has_valueof_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_method(enum_cls->get_type()));
  EXPECT_TRUE(has_values_field(enum_cls->get_type()));
}
